### PR TITLE
Added support for material accounting in schedulers

### PR DESCRIPTION
**Pull request by vanoha**
- added landscape to schedulers
- added landscape to momentum timeline
- tested material accounting in the schedulers
- built universal test for accounting materials in schedulers and timelines

### CODE REVIEW

File: sampo/scheduler/genetic/operators.py [hunk 1]
```diff
@@ -194,9 +205,10 @@ def generate_chromosome(wg: WorkGraph, contractors: list[Contractor], index2node
     :return: chromosome
     """
 
-    def randomized_init() -> ChromosomeType:
+    def randomized_init(_landscape: LandscapeConfiguration = LandscapeConfiguration()) -> ChromosomeType:
```
Comment (StannisMod) [inline sampo/scheduler/genetic/operators.py:208]: this is an inner function, so it inherits 'landscape' parameter

File: sampo/scheduler/timeline/just_in_time_timeline.py [hunk 2]
```diff
@@ -72,7 +77,15 @@ def find_min_start_time_with_additional(self, node: GraphNode,
                 needed_count -= offer_count
                 ind -= 1
 
-        c_st = max(max_agent_time, max_parent_time, max_neighbor_time)
+        c_st = max(max_agent_time, max_parent_time)
```
Comment (StannisMod) [inline sampo/scheduler/timeline/just_in_time_timeline.py:80]: where is max_neighbor_time??

File: sampo/scheduler/timeline/material_timeline.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+
+        grabbed = 0
+
+        cur_start_time = deadline
+        going_right = False
+
+        for material in materials:
+            try:
+                depot = self._find_best_supply(material.name, material.count)
+            except Exception as e:
```
Comment (StannisMod) [inline sampo/scheduler/timeline/material_timeline.py:181]: exception-driven flow, performance impacts

File: sampo/scheduler/topological/base.py [hunk 4]
```diff
@@ -51,9 +55,11 @@ class RandomizedTopologicalScheduler(TopologicalScheduler):
     """
     Scheduler, that represent 'WorkGraph' in topological order with random
     """
-    def __init__(self, work_estimator: Optional[WorkTimeEstimator or None] = None,
-                 random_seed: Optional[int] = None):
-        super().__init__(work_estimator=work_estimator)
+    def __init__(self,
+                 work_estimator: Optional[WorkTimeEstimator or None] = None,
+                 random_seed: Optional[int] = None,
+                 landscape: LandscapeConfiguration = LandscapeConfiguration()):
```
Comment (StannisMod) [inline sampo/scheduler/topological/base.py:61]: landscape configuration in Scheduler init...

File: sampo/schemas/graph.py [hunk 5]
```diff
@@ -354,6 +368,7 @@ def _to_adj_matrix(self) -> tuple[list[GraphNode], dok_matrix, dict[str, GraphNo
         }
         id2node: dict[str, GraphNode] = {node.id: node for node in node2ind.keys()}
         adj_mx = dok_matrix((len(node2ind), len(node2ind)), dtype=np.short)
+        weight = 0
```
Comment (StannisMod) [inline sampo/schemas/graph.py:371]: why? 'weight' is always initialized when accessed

File: sampo/schemas/scheduled_work.py [hunk 6]
```diff
@@ -32,7 +33,7 @@ def __init__(self,
                  start_end_time: tuple[Time, Time],
                  workers: list[Worker],
                  contractor: Contractor | str,
-                 equipments: list[Equipment] | None = None,
+                 equipments: list[MaterialDelivery] | None = None,
```
Comment (StannisMod) [inline sampo/schemas/scheduled_work.py:36]: equipments with MaterialDelivery type?

File: tests/conftest.py [hunk 7]
```diff
@@ -35,20 +38,36 @@ def setup_rand() -> Random:
     return Random(231)
 
 
+@fixture
+def setup_landscape():
```
Comment (StannisMod) [inline tests/conftest.py:42]: rename with 'setup_landscape_one_holder'
the same is for next fixture - 'with' don't need here

File: tests/conftest.py [hunk 8]
```diff
@@ -35,20 +38,36 @@ def setup_rand() -> Random:
...
     return SimpleSynthetic(setup_rand)
 
 
 @fixture(params=[(graph_type, lag) for lag in [True, False]
                  for graph_type in ['manual',
-                                    'small plain synthetic', 'big plain synthetic',]],
-                                    # 'small advanced synthetic', 'big advanced synthetic']],
+                                    ]],
+         # 'small plain synthetic', 'big plain synthetic']],
+         # 'small advanced synthetic', 'big advanced synthetic']],
```
Comment (StannisMod) [inline tests/conftest.py:65]: REVERT. 
1) You should't disable synthetic graphs in tests
2) the code style in this place is correct

File: tests/conftest.py [hunk 9]
```diff
@@ -35,20 +38,36 @@ def setup_rand() -> Random:
...
          ids=[f'Graph: {graph_type}, LAG_OPT={lag_opt}'
               for lag_opt in [True, False]
               for graph_type in ['manual',
-                                 'small plain synthetic', 'big plain synthetic',]])
-                                 # 'small advanced synthetic', 'big advanced synthetic']])
+                                 ]])
+# 'small advanced synthetic', 'big advanced synthetic']])
```
Comment (StannisMod) [inline tests/conftest.py:70]: the same here

File: tests/scheduler/material_scheduling_test.py [hunk 10]
```diff
@@ -0,0 +1,39 @@
+import pytest
+
+from sampo.scheduler.heft.base import HEFTBetweenScheduler
+from sampo.scheduler.heft.base import HEFTScheduler
+from sampo.utilities.validation import validate_schedule
+
+
+def test_just_in_time_scheduling_with_materials(setup_default_schedules):
```
Comment (StannisMod) [inline tests/scheduler/material_scheduling_test.py:8]: you use setup_default_schedules, but not use schedules?)

**Review by StannisMod — CHANGES_REQUESTED** [review]
Critical bugs created, changes required.

File: sampo/schemas/graph.py (hunk 5 above)
Comment (vanoha) [reply sampo/schemas/graph.py:371]: yes, but at the same time no
initialization of 'weight' is in the 'try' zone, but accesse to 'weight' is out of 'try'
python finds this variable, but writing it this way makes more sense

File: sampo/scheduler/genetic/operators.py [hunk 11]
```diff
@@ -235,7 +247,7 @@ def is_chromosome_correct(chromosome: ChromosomeType,
     :return:
     """
     return is_chromosome_order_correct(chromosome, parents) and \
-           is_chromosome_contractors_correct(chromosome, node_indices)
+        is_chromosome_contractors_correct(chromosome, node_indices)
```
Comment (StannisMod) [inline sampo/scheduler/genetic/operators.py:250]: the same, alignment... i suppose that old alignment is better to read, isn't it?

File: sampo/scheduler/timeline/material_timeline.py [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+        finish_time = max([b[1] for b in batch_processing])
+        deliveries.extend([b[0] for b in batch_processing])
+
+        return deliveries, start_time, finish_time
+
+    def _find_best_supply(self, material: str, count: int) -> str:
+        # TODO Make better algorithm
+        # Return the first depot that can supply given materials
+        if len(self._resource_sources) == 0:
+            raise NoAvailableResources(f"Schedule can not be built. No available resource {material}")
```
Comment (StannisMod) [inline sampo/scheduler/timeline/material_timeline.py:67]: In this place there are not available resource sources, not resource

File: tests/conftest.py [hunk 13]
```diff
@@ -137,34 +164,33 @@ def setup_scheduler_parameters(request, setup_wg) -> tuple[WorkGraph, list[Contr
...
-    return setup_wg, contractors
+    return setup_wg, contractors, setup_landscape_many_holders
 
 
 @fixture
 def setup_default_schedules(setup_scheduler_parameters):
     work_estimator: Optional[WorkTimeEstimator] = None
 
-    setup_wg, setup_contractors = setup_scheduler_parameters
+    setup_wg, setup_contractors, landscape = setup_scheduler_parameters
```
Comment (StannisMod) [inline tests/conftest.py:174]: Input parameters from fixtures should be named from `setup_`

File: tests/scheduler/genetic/converter_test.py [hunk 14]
```diff
@@ -6,10 +6,10 @@
 from sampo.utilities.validation import validate_schedule
 
 
-def test_convert_schedule_to_chromosome(setup_toolbox):
+def test_convert_schedule_to_chromosome(setup_toolbox, setup_landscape_many_holders):
```
Comment (StannisMod) [inline tests/scheduler/genetic/converter_test.py:9]: `setup_landscape_many_holders` should come from `setup_scheduler_parameters`, and from `setup_toolbox` too.
Here you just duplicating tests.

File: tests/scheduler/genetic/fixtures.py [hunk 15]
```diff
@@ -132,7 +135,7 @@ def create_toolbox(wg: WorkGraph,
 
 @fixture
 def setup_toolbox(setup_default_schedules) -> tuple:
-    (setup_wg, setup_contractors), setup_default_schedules = setup_default_schedules
+    (setup_wg, setup_contractors, landscape), setup_default_schedules = setup_default_schedules
```
Comment (StannisMod) [inline tests/scheduler/genetic/fixtures.py:138]: In this place too, from `setup_`

**Review by StannisMod — CHANGES_REQUESTED** [review]
Some cosmetic things required :)

**Review by StannisMod — APPROVED** [review]
Approved

**Pull request merged**